crates/turborepo-ffi/src/scm/git.rs

```diff
@@ -0,0 +1,256 @@
+use std::{collections::HashSet, path::PathBuf};
+
+use anyhow::{anyhow, Result};
+use dunce::canonicalize as fs_canonicalize;
+use git2::{DiffFormat, DiffOptions, Oid, Repository};
+
+/// Finds the changed files in a repository between index and working directory
+/// (unstaged changes) and between two commits.
+///
+/// # Arguments
+///
+/// * `repo_root`: The root of the repository.
+/// * `commit_range`: If Some, the range of commits that should be searched for
+///   changes
+/// * `include_untracked`: If true, untracked files will be included in the
+///   result, i.e. files not yet in git.
+/// * `relative_to`: If Some, only the results relative to this path will be
+///   returned
+///
+/// returns: Result<HashSet<String, RandomState>, Error>
+pub fn changed_files(
+    repo_root: PathBuf,
+    commit_range: Option<(&str, &str)>,
+    include_untracked: bool,
+    relative_to: Option<&str>,
+) -> Result<HashSet<String>> {
+    let repo = Repository::open(repo_root)?;
+    let mut files = HashSet::new();
+    add_changed_files_from_unstaged_changes(&repo, &mut files, relative_to, include_untracked)?;
+
+    if let Some((from_commit, to_commit)) = commit_range {
+        add_changed_files_from_commits(&repo, &mut files, relative_to, from_commit, to_commit)?;
+    }
+
+    Ok(files)
+}
+
+fn add_changed_files_from_unstaged_changes(
+    repo: &Repository,
+    files: &mut HashSet<String>,
+    relative_to: Option<&str>,
+    include_untracked: bool,
+) -> Result<()> {
+    let mut options = DiffOptions::new();
+    options.include_untracked(include_untracked);
+    options.recurse_untracked_dirs(include_untracked);
+    relative_to.map(|relative_to| options.pathspec(relative_to));
+
+    let diff = repo.diff_index_to_workdir(None, Some(&mut options))?;
+
+    for delta in diff.deltas() {
+        let file = delta.old_file();
+        if let Some(path) = file.path() {
+            // NOTE: In the original Go code, `Rel` works even if the base path is not a
+            // prefix of the original path. In Rust, `strip_prefix` returns an
+            // error if the base path is not a prefix of the original path.
+            // However since we're passing a pathspec to `git2` we know that the
+            // base path is a prefix of the original path.
+            let path = relative_to.map_or(path, |relative_to| {
+                path.strip_prefix(relative_to).unwrap_or(path)
+            });
+            files.insert(path.to_string_lossy().to_string());
```

Comment:
I don't think we should be silently failing if we get a non-UTF8 path. If we don't fail here, we're going to get a weirder failure up the call stack.
```suggestion
            files.insert(path.to_str().ok_or(anyhow!("non utf8 path"))?.to_string());
```

Reply:
Yeah this is something we should standardize as a whole. `to_string_lossy` kinda snuck into our codebase and now it's in a few places.